vnext/Microsoft.ReactNative/Utils/TextHitTestUtils.cpp

```diff
@@ -0,0 +1,328 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+
+#include "TextHitTestUtils.h"
+
+#include <UI.Xaml.Documents.h>
+
+namespace Microsoft::ReactNative {
+
+struct BidirectionalTextBoundary {
+  xaml::Documents::TextPointer ltrPointer{nullptr};
+  xaml::Documents::TextPointer rtlPointer{nullptr};
+  winrt::Rect ltrRect;
+  winrt::Rect rtlRect;
+};
+
+/**
+ * Finds the character closest to X = 0 for the first line. I.e., the
+ * leftmost character when flow direction is left-to-right.
+ * ┌─────> X ↓
+ * │ (x,y)  ┌─<─┐<┌─>─┐
+ * │ ┌──────┘   └─┘   │
+ * │ │                │
+ * ↓ │     ┌──────────┘
+ * Y └─────┘
+ */
+static BidirectionalTextBoundary FindStartRTLBoundary(
+    const xaml::Documents::TextPointer &start,
+    const xaml::Documents::TextPointer &end,
+    const winrt::Rect &target) {
+  auto textPointer = start;
+  auto L = start.Offset();
+  auto R = end.Offset();
+  while (L <= R) {
+    const auto m = /* floor */ (L + R) / 2;
+    const auto relativeOffset = m - textPointer.Offset();
+    textPointer = textPointer.GetPositionAtOffset(relativeOffset, xaml::Documents::LogicalDirection::Forward);
+    const auto rect = textPointer.GetCharacterRect(xaml::Documents::LogicalDirection::Forward);
+    // If median character is on different line, boundary is before median
+    if (rect.Y != target.Y) {
+      R = m - 1;
```

Comment:
these are `float` values, avoid equality comparisons wherever possible (or use an epsilon)

Reply:
I can add an epsilon here, but I believe these are known to be exact because all characters in a given line are given the same Y value, even when they have tall ascenders.